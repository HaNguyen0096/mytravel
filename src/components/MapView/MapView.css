.mapImage{
  height: 50px;
  width: auto;
}