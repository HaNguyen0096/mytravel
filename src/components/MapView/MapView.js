import React, { useState } from 'react'
import ReactMapGL, {Marker, Popup} from 'react-map-gl'
import ViewLogBtn from '../ViewLogBtn/ViewLogBtn'
import AddLog from '../AddLog/AddLog'
import './MapView.css'
import 'mapbox-gl/dist/mapbox-gl.css'

function MapView(props) {
  const [showPopup, setShowPopup] = useState({})
  const [addLogLocation, setAddLogLocation] = useState(null)
  const [viewport, setViewport] = useState({
    width: '100vw',
    height: '100vh',
    latitude: 42,
    longitude: -71,
    zoom: 8
  });

  const showAddMarkerPopup = (event) => {
    const [ longitude, latitude ] = event.lngLat;
    setAddLogLocation({
      latitude,
      longitude,
    });
  };

  const {logs, viewPath} = props
  const checkPath = viewPath.substring(1,7)

  return (
    <ReactMapGL
      {...viewport}
      mapboxApiAccessToken={process.env.REACT_APP_MAPBOX_TOKEN}
      onViewportChange={setViewport}
      onDblClick={showAddMarkerPopup}
    >
      {
        logs.map(log => (
          <React.Fragment key={log.id}>
            <Marker   
              key={log.id}         
              latitude={Number(log.latitude)} 
              longitude={Number(log.longitude)} 
              offsetLeft={-12}
              offsetTop={-24}
            >
              <div
                onClick={() => setShowPopup({
                  [log.id]:true,
                })}
              >
                <svg 
                  className='marker'
                  style={{
                    height: `24px`,
                    width: `24px`,
                  }}
                  stroke='currentColor' 
                  strokeWidth='2' fill='none' 
                  strokeLinecap='round' 
                  strokeLinejoin='round' 
                >
                    <path d='M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z'></path>
                    <circle cx='12' cy='10' r='3'></circle>
                </svg>
              </div>
            </Marker>
            {
              showPopup[log.id] ? (
                <Popup
                  latitude={Number(log.latitude)}
                  longitude={Number(log.longitude)}
                  closeButton={true}
                  closeOnClick={false}
                  dynamicPosition={true}
                  onClose={() => setShowPopup({})}
                  anchor='top' >
                  <div className='popup'>
                    <div className='mapLogTitle'>
                      <h3>{log.title}</h3>
                    </div>
                    <div ></div>
                    {log.image && <img className='mapImage' src={log.image} alt={log.title} />}
                    <div className='Btn'>
                      <ViewLogBtn id={log.id}/>
                    </div>
                  </div>
                </Popup>
              ) : null
            }
          </React.Fragment>
        ))       
      }

      {
        (checkPath === 'mylogs') && addLogLocation ? (
          <div>
          <Marker
            latitude={addLogLocation.latitude}
            longitude={addLogLocation.longitude}
          >
            <div>
              <svg 
                className='marker'
                style={{
                  height: `24px`,
                  width: `24px`,
                }}
                stroke='currentColor' 
                strokeWidth='2' fill='none' 
                strokeLinecap='round' 
                strokeLinejoin='round' 
              >
                <path d='M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z'></path>
                <circle cx='12' cy='10' r='3'></circle>
              </svg>
            </div>
          </Marker>
          <Popup
            latitude={addLogLocation.latitude}
            longitude={addLogLocation.longitude}
            closeButton={true}
            closeOnClick={false}
            dynamicPosition={true}
            onClose={() => setAddLogLocation(null)}
            anchor='top' >
            <div className='popup'>
              <AddLog onClose={() => {
                setAddLogLocation(null);
              }} location={addLogLocation} />
            </div>
          </Popup>
          </div>
        ) : null
      }

    </ReactMapGL>
  );
}


export default MapView;