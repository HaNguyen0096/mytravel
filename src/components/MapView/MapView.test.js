import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import MapView from './MapView'

describe('MapView component', () => {
  const viewPath = 'MapView'
  const testLogs = [
    {
      id: 1,
      latitude: '42.3554',
      longitude: '-71.0640',
      title: 'Boston Common',
      description: 'This is a beautiful place',
      image: 'https://en.wikipedia.org/wiki/File:Aerial_View_Parkman_Bandstand_at_Boston_Common.jpg',
      rating: 10,
      visited_day: '2020-03-18T07:55:12.825Z',
      public: false,
      date_created: '2020-03-18T07:55:12.825Z',
    },
    {
      id: 2,
      latitude: '48.8584',
      longitude: '2.2945',
      title: 'Eiffel Tower',
      description: 'The tower is so tall',
      image: 'https://upload.wikimedia.org/wikipedia/commons/8/85/Tour_Eiffel_Wikimedia_Commons_%28cropped%29.jpg',
      rating: 10,
      visited_day: '2020-03-18T07:55:12.825Z',
      public: true,
      date_created: '2020-03-18T07:55:12.825Z'
    },
  ]
  it('renders MapView without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
                      <MapView viewPath={viewPath} logs={testLogs}/>
                    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI for MapView as expected', () => {
    const tree = renderer
      .create(<BrowserRouter>
                <MapView viewPath={viewPath} logs={testLogs}/>
              </BrowserRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });

  it('renders MapView by default', () => {
    const wrapper = shallow(<MapView  viewPath={viewPath} logs={testLogs}/>)
    expect(toJson(wrapper)).toMatchSnapshot()
  });

})