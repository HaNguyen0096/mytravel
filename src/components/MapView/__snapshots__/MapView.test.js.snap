// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapView component renders MapView by default 1`] = `
<InteractiveMap
  altitude={1.5}
  asyncRender={false}
  attributionControl={true}
  bearing={0}
  className=""
  clickRadius={0}
  container={<body />}
  disableTokenWarning={false}
  doubleClickZoom={true}
  dragPan={true}
  dragRotate={true}
  getCursor={[Function]}
  height="100vh"
  keyboard={true}
  latitude={42}
  longitude={-71}
  mapOptions={Object {}}
  mapStyle="mapbox://styles/mapbox/light-v8"
  mapboxApiAccessToken="pk.eyJ1IjoiaGFuZ3V5ZW45NiIsImEiOiJjazd2Y3B1MjgwYzk0M2VwY3NxdXM3cDhoIn0.x7HcFQ5ULW_x2mLUqgvs4w"
  mapboxApiUrl="https://api.mapbox.com"
  maxPitch={60}
  maxZoom={24}
  minPitch={0}
  minZoom={0}
  onClick={null}
  onContextMenu={[Function]}
  onDblClick={[Function]}
  onError={[Function]}
  onHover={null}
  onLoad={[Function]}
  onNativeClick={null}
  onResize={[Function]}
  onStateChange={[Function]}
  onTransitionEnd={[Function]}
  onTransitionInterrupt={[Function]}
  onTransitionStart={[Function]}
  onViewStateChange={null}
  onViewportChange={[Function]}
  pitch={0}
  preserveDrawingBuffer={false}
  preventStyleDiffing={false}
  reuseMaps={false}
  scrollZoom={true}
  style={null}
  touchAction="none"
  touchRotate={false}
  touchZoom={true}
  transitionDuration={0}
  transitionEasing={[Function]}
  transitionInterpolator={
    LinearInterpolator {
      "around": undefined,
      "propNames": Array [
        "longitude",
        "latitude",
        "zoom",
        "bearing",
        "pitch",
      ],
    }
  }
  transitionInterruption={1}
  visibilityConstraints={
    Object {
      "maxPitch": 60,
      "maxZoom": 24,
      "minPitch": 0,
      "minZoom": 0,
    }
  }
  visible={true}
  width="100vw"
  zoom={8}
>
  <Marker
    captureClick={true}
    captureDoubleClick={true}
    captureDrag={true}
    captureScroll={false}
    className=""
    draggable={false}
    key="1"
    latitude={42.3554}
    longitude={-71.064}
    offsetLeft={-12}
    offsetTop={-24}
  >
    <div
      onClick={[Function]}
    >
      <svg
        className="marker"
        fill="none"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          Object {
            "height": "24px",
            "width": "24px",
          }
        }
      >
        <path
          d="M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z"
        />
        <circle
          cx="12"
          cy="10"
          r="3"
        />
      </svg>
    </div>
  </Marker>
  <Marker
    captureClick={true}
    captureDoubleClick={true}
    captureDrag={true}
    captureScroll={false}
    className=""
    draggable={false}
    key="2"
    latitude={48.8584}
    longitude={2.2945}
    offsetLeft={-12}
    offsetTop={-24}
  >
    <div
      onClick={[Function]}
    >
      <svg
        className="marker"
        fill="none"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          Object {
            "height": "24px",
            "width": "24px",
          }
        }
      >
        <path
          d="M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z"
        />
        <circle
          cx="12"
          cy="10"
          r="3"
        />
      </svg>
    </div>
  </Marker>
</InteractiveMap>
`;

exports[`MapView component renders the UI for MapView as expected 1`] = `
<div
  style={
    Object {
      "cursor": "grab",
      "height": 0,
      "position": "relative",
      "width": 0,
    }
  }
>
  <div
    style={
      Object {
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      className=""
      style={
        Object {
          "height": "100%",
          "overflow": "hidden",
          "position": "absolute",
          "visibility": "inherit",
          "width": "100%",
        }
      }
    />
    <div
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    />
  </div>
</div>
`;
